### PR TITLE
#1109F Trigger/Function for deleting review question assignments

### PR DESCRIPTION
**Pull request by CarlosNZ**
Required for [front-end issue #1109](https://github.com/openmsupply/application-manager-web-app/issues/1109) (PR https://github.com/openmsupply/application-manager-web-app/pull/1128)

The front-end had no knowledge of the state of existing review question assignments for items on the "Assignment" tab, so rather than write a bunch of additional front-end queries to fetch this info and then process it, it is simpler to leave the question_assignment handling to database triggers & functions.

There was a slight headache, in that the `delete_question_assignments` function's main query didn't work when "assigned_sections" was an empty array, so had to have slightly different logic for that case. Basically:
- trigger function runs when "assigned_sections" on "review_assignment" changes (either from the front-end, or reset via the existing review_assignment trigger)
- If sections remain assigned, this function deletes question assignments relating to the remaining sections
- If no sections remain (we check for "status = AVAILABLE" instead), ALL review question assignments are deleted

Also here:
- Added a generated column to "review_question_assignment" to show the template section code
- Update migration script (with the above plus a couple of refinements to existing)
- Update documentation regarding data/schema migration

### CODE REVIEW

**Comment by nmadruga** [comment]
> * If no sections remain (we check for "status = AVAILABLE" instead), ALL review question assignments are deleted

I was thinking about this one... This will change slightly the logic that we added previously. Basically we wanted to keep the state of a previous review done by a reviewer when they get un-assigned from a review. Just in case it is re-assigned to them again iit would be keeping their previous reviewResponses.

I was lookiing for where this was explained in the Wiki, but could only find this:

> [Re-Assignment](https://github.com/openmsupply/application-manager-web-app/wiki/Assignment-Process#re-assignment)
> A GraphQL mutation reassignReviewAssignment runs on the front-end when one assigner:
> 
> selects re-assign in the drop down
> selects the new reviewer in drop down for a section (or all sections).
> review_assignment.trigger for previous assignment is set to ON_REVIEW_UNASSIGN.
> review_assignment.trigger for new assignment is set to ON_REVIEW_REASSIGN.
> Changes status of previous review_assignment to Available and flag is_locked to True.
> Changes status of new review_assignment to Assigned.
> Sets assignerId of current user in the new assignment.
> Adds review_question_assignment (for each question on the section assigned) to review_assignment.
> Other updates done on the server-side:
> If an existing review was connected to previous assignment will change the review.status from Draft to Discontinued.
> **If an existing review was connected to new assignment will change the review.status from Discontinued to Draft.**

**Review by nmadruga — COMMENTED** [review]
Looks good. I also tested using the snapshot `__just_before_assignment`, and saw the questions being correctly removed from one assignment and created on the new assignment to another user. And I was thinking about the idea of `DISCUTINUED` status. It seems it's going to be very hard to keep both logic. IIf we want to do half re-assignment I don't think we can also keep the ability of **continuing** a previously assigned (and unassigned) review questions. So I'm happy with just changing the docs on the front end so we remove the logic for DISCONTINUE a review.

**Comment by CarlosNZ** [comment]
Okay, have added significant changes to update some of the functions to handle the removal of review_question_assignments and add migration code. Can be tested with https://github.com/openmsupply/application-manager-web-app/pull/1128

File: database/buildSchema/34_review_status_history.sql [hunk 1]
```diff
@@ -92,18 +92,42 @@ CREATE FUNCTION public.submitted_assigned_questions_count (app_id int, stage_id
...
+        FROM
+            TEMPLATE
+        WHERE
+            id = (
+                SELECT
+                    template_id
+                FROM
+                    application
+                WHERE
+                    id = 236))
```
Comment (nmadruga) [inline database/buildSchema/34_review_status_history.sql:129]: Something left behind here?

File: database/migration/migrateData.ts [hunk 2]
```diff
@@ -57,24 +58,76 @@ const migrateData = async () => {
...
+    AND ra.stage_id = $2
+    AND ra.level_number = $3
+    AND ra.status = 'ASSIGNED'
+    AND te.category = 'QUESTION'
+    AND rsh.status = 'SUBMITTED'
+    AND te.template_code = (
+        SELECT code FROM TEMPLATE
+        WHERE id = (
+          SELECT template_id FROM application
+          WHERE id = 236))
```
Comment (nmadruga) [inline database/migration/migrateData.ts:126]: Again something that might be left behind?

File: plugins/action_update_review_status/src/databaseMethods.ts [hunk 3]
```diff
@@ -55,9 +55,17 @@ const databaseMethods = (DBConnect: any) => ({
...
-    WHERE review_assignment_id = $1
+    SELECT DISTINCT(te.id) AS "templateElementId"
+    FROM template_element te JOIN template_section ts
+    ON te.section_id = ts.id
+    JOIN (
+    		SELECT id, application_id, template_id, UNNEST(assigned_sections) as section_code
+    		FROM review_assignment
+    	) ra 
+    ON ra.template_id = ts.template_id  
+    WHERE ra.id = 28
```
Comment (nmadruga) [inline plugins/action_update_review_status/src/databaseMethods.ts:66]: Another hardcoded id

**Review by nmadruga — CHANGES_REQUESTED** [review]
Some hardcoded ids, I'm confused with what needs to be included and what is temp code that you were using to test with this particular example

**Comment by CarlosNZ** [comment]
> Some hardcoded ids, I'm confused with what needs to be included and what is temp code that you were using to test with this particular example

Ooops, sorry. 😳

Fixed now dcdba919

**Review by nmadruga — APPROVED** [review]

**Pull request merged**